zipline/data/data_portal.py

```diff
@@ -311,6 +323,146 @@ def _get_daily_data(self, asset_int, column, dt):
+                # 1) the history window ends at midnight utc.
+                # 2) the last desired day of the window is after the
+                # last trading day, use daily data for the whole range.
+                data.append(self._get_daily_window_for_sid(
+                    sid,
+                    field_to_use,
+                    days_for_window,
+                    extra_slot=False
+                ))
+            else:
+                # for the last day of the desired window, use minute
+                # data and aggregate it.
+                all_minutes_for_day = self.env.market_minutes_for_day(
+                    pd.Timestamp(day))
+
+                last_minute_idx = all_minutes_for_day.searchsorted(end_dt)
+
+                # these are the minutes for the partial day
+                minutes_for_partial_day =\
+                    all_minutes_for_day[0:(last_minute_idx + 1)]
+
+                daily_data = self._get_daily_window_for_sid(
+                    sid,
+                    field_to_use,
+                    days_for_window[0:-1]
+                )
+
+                minute_data = self._get_minute_window_for_sid(
+                    sid,
+                    field_to_use,
+                    minutes_for_partial_day
+                )
+
+                if field_to_use == 'volume':
+                    minute_value = np.sum(minute_data)
+                elif field_to_use == 'open':
+                    minute_value = minute_data[0]
+                elif field_to_use == 'close':
+                    minute_value = minute_data[-1]
+                elif field_to_use == 'high':
+                    minute_value = np.amax(minute_data)
+                elif field_to_use == 'low':
+                    minute_value = np.amin(minute_data)
+
+                # append the partial day.
+                daily_data[-1] = minute_value
+
+                data.append(daily_data)
+
+        return pd.DataFrame(np.array(data).T,
+                            index=days_for_window,
+                            columns=sids)
+
+    def _get_history_minute_window(self, sids, end_dt, bar_count,
+                                   field_to_use):
+        """
+        Internal method that returns a dataframe containing history bars
+        of minute frequency for the given sids.
+        """
+        # get all the minutes for this window
+        minutes_for_window = self.env.market_minute_window(
+            end_dt, bar_count, step=-1)[::-1]
+
+        # but then cut it down to only the minutes after
+        # FIRST_TRADING_MINUTE
+        modified_minutes_for_window = minutes_for_window[
+            minutes_for_window.slice_indexer(FIRST_TRADING_MINUTE)]
+
+        modified_minutes_length = len(modified_minutes_for_window)
+
+        if modified_minutes_length == 0:
+            raise ValueError("Cannot calculate history window that ends"
+                             "before 2002-01-02 14:31 UTC!")
```

Comment:
Yeah, there's a task on the punchlist for dealing with all the 1/1/2002 stuff that is currently in dataportal (there's a bunch of stuff at the top of the file that needs moving too).